Detectors/CPV/calib/src/NoiseCalibrator.cxx

```diff
@@ -0,0 +1,150 @@
+NoiseCalibData::NoiseCalibData()
+{
+  auto& cpvParams = o2::cpv::CPVSimParams::Instance();
+  mNoiseThreshold = cpvParams.mNoiseClbThreshold;
+  for (int i = 0; i < Geometry::kNCHANNELS; i++) {
+    mOccupancyMap.push_back(0);
+  }
+}
+//_____________________________________________________________________________
+void NoiseCalibData::fill(const gsl::span<const Digit> digits)
+{
+  for (auto& dig : digits) {
+    if (dig.getAmplitude() > mNoiseThreshold) {
+      mOccupancyMap[dig.getAbsId()]++;
+    }
+  }
+  mNEvents++;
+}
+//_____________________________________________________________________________
+void NoiseCalibData::merge(const NoiseCalibData* prev)
+{
+  for (int i = 0; i < Geometry::kNCHANNELS; i++) {
+    mOccupancyMap[i] += prev->mOccupancyMap[i];
+  }
+  mNEvents += prev->mNEvents;
+  LOG(info) << "Merged TimeSlot with previous one. Now we have " << mNEvents << " events.";
+}
+//_____________________________________________________________________________
+void NoiseCalibData::print()
+{
+  LOG(info) << "NoiseCalibData::mNEvents = " << mNEvents;
+}
+//_____________________________________________________________________________
+// NoiseCalibrator
+//_____________________________________________________________________________
+NoiseCalibrator::NoiseCalibrator()
+{
+  LOG(info) << "NoiseCalibrator::NoiseCalibrator() : "
+            << "Noise calibrator created!";
+  auto& cpvParams = o2::cpv::CPVSimParams::Instance();
+  mMinEvents = cpvParams.mNoiseClbMinEvents;
+  mToleratedChannelEfficiencyLow = cpvParams.mNoiseClbToleratedChannelEfficiencyLow;
+  mToleratedChannelEfficiencyHigh = cpvParams.mNoiseClbToleratedChannelEfficiencyHigh;
+  mNoiseFrequencyCriteria = cpvParams.mNoiseClbFrequencyCriteria;
+}
+//_____________________________________________________________________________
+void NoiseCalibrator::initOutput()
+{
+  LOG(info) << "NoiseCalibrator::initOutput() : output vectors cleared";
+  mCcdbInfoBadChannelMapVec.clear();
+  mBadChannelMapVec.clear();
+}
+//_____________________________________________________________________________
+void NoiseCalibrator::finalizeSlot(NoiseTimeSlot& slot)
+{
+  NoiseCalibData* calibData = slot.getContainer();
+  LOG(info) << "NoiseCalibrator::finalizeSlot() : finalizing slot "
+            << slot.getTFStart() << " <= TF <= " << slot.getTFEnd() << " with " << calibData->mNEvents << " events.";
+  o2::cpv::BadChannelMap* badMap = new o2::cpv::BadChannelMap();
+  bool badMapBool[Geometry::kNCHANNELS] = {};
+
+  // handle data from pedestal run first
+  // check pedestal efficiencies
+  if (mPedEfficiencies) {
+    LOG(info) << "NoiseCalibrator::finalizeSlot() : checking ped efficiencies";
+    for (int i = 0; i < Geometry::kNCHANNELS; i++) {
+      badMapBool[i] = false;
+      if (mPedEfficiencies.get()->at(i) > mToleratedChannelEfficiencyHigh ||
+          mPedEfficiencies.get()->at(i) < mToleratedChannelEfficiencyLow) {
```

Comment:
do you really need to use `at(i)` which does a range check? `[i]` is faster.

Reply:
That I didn't know. Thanks for pointing!